[feature/message-queue] Message Queue to allow flexible handling of issues and messages

felix-schwarz opened the pull request:
## Description
Introduces a new message queue mechanism that allows fine-grained routing, decoupled and tailored presentation, handling and automatic resolution of issues and messages. Works across the app and extensions like the File Provider.

## Related Issue
#362 #505 #513 #585 https://github.com/owncloud/ios-sdk/pull/63

## Screenshots (if appropriate):
### Dismissable card in-app
Issues from actions started from within the app pop up as cards. They can be dismissed by tapping in the area above the card to deal with them later.
![Bildschirmfoto 2020-03-27 um 16 28 41](https://user-images.githubusercontent.com/639669/77773276-aa482600-7049-11ea-91ae-0dab9fc6ff76.png)

### Issue Handling via Status tab
Issues and messages the user didn't take action on yet are listed in the Status tab. Identical issues will eventually be grouped and offers the option to apply a choice (like "Retry" or "Cancel") to all of them at once. 
![Bildschirmfoto 2020-03-27 um 16 27 09](https://user-images.githubusercontent.com/639669/77773281-ab795300-7049-11ea-871e-1f7ab1e99742.png)

### Notifications from File Provider
Errors and issues originating from actions started through the Files Provider / Files app are shown as notifications.
![Bildschirmfoto 2020-03-27 um 16 26 33](https://user-images.githubusercontent.com/639669/77773282-ab795300-7049-11ea-8244-a120edaa313c.png)

### Responding to notifications
Issues can be dealt with directly in the notifications:
<img width="559" alt="Bildschirmfoto 2020-03-31 um 00 49 22" src="https://user-images.githubusercontent.com/639669/77969394-f0bbb000-72e9-11ea-8204-bebd1dcded0d.png">

Tapping notifications shows the message in the context of the originating account - or, if not possible - in the context of another account with an added header (pictured below):
<img width="1376" alt="Bildschirmfoto 2020-04-27 um 23 35 21" src="https://user-images.githubusercontent.com/639669/80423383-3e730a80-88e0-11ea-97b5-eac728dab292.png">

### Grouping of similar messages
If there are two or more similar messages, only the oldest one is presented directly, while options to apply the choice to all - or to show them all - are added below the issue:
<img width="1376" alt="Bildschirmfoto 2020-05-26 um 22 24 09" src="https://user-images.githubusercontent.com/639669/82947175-e5120000-9f9f-11ea-9d7b-e0dc4a17038c.png">

Turning on `Apply to all` adds an additional badge to clarify the choice will apply to more than one message:
<img width="1376" alt="Bildschirmfoto 2020-05-26 um 22 24 38" src="https://user-images.githubusercontent.com/639669/82947309-1c80ac80-9fa0-11ea-9bd2-9bad2e6c7782.png">

Tapping `Show all` shows all of the grouped messages and allows to respond to them individually:
<img width="1332" alt="Bildschirmfoto 2020-05-26 um 22 24 25" src="https://user-images.githubusercontent.com/639669/82947326-273b4180-9fa0-11ea-96b5-3b4f2d8ecd7c.png">

### Inline information
If there's a message related to an item or sync action, a ⚠️ warning sign appears next to it. Tapping it then brings up the message for the user to act on:
| In the file list | In the status tab |
| --- | --- |
| <img width="559" alt="Bildschirmfoto 2020-06-17 um 22 57 28" src="https://user-images.githubusercontent.com/639669/84950081-93681b80-b0ee-11ea-82fa-59b3726ec8a1.png"> | <img width="559" alt="Bildschirmfoto 2020-06-17 um 22 57 37" src="https://user-images.githubusercontent.com/639669/84950069-8f3bfe00-b0ee-11ea-8495-9eaa1799b60b.png"> |
| After tapping: | After tapping: |
| <img width="559" alt="Bildschirmfoto 2020-06-17 um 22 57 34" src="https://user-images.githubusercontent.com/639669/84950206-d4f8c680-b0ee-11ea-81f3-58e3af74ab94.png"> | <img width="559" alt="Bildschirmfoto 2020-06-17 um 22 57 40" src="https://user-images.githubusercontent.com/639669/84950202-d3c79980-b0ee-11ea-976d-7b611a3deb42.png"> |

### "All done" 
Displays an "All done" message when there are no messages or activities:
![Bildschirmfoto 2020-06-25 um 15 53 48](https://user-images.githubusercontent.com/639669/85733669-aab19500-b6fc-11ea-8c27-41c00a470fa6.png)

### Per-account message count badge
In the bookmark list:
<img width="559" alt="Bildschirmfoto 2020-03-31 um 00 49 47" src="https://user-images.githubusercontent.com/639669/77969493-1ba60400-72ea-11ea-999d-4e04121b44ad.png">

### App Icon
Total message count badge on the app icon:
<img width="96" alt="Bildschirmfoto 2020-03-31 um 00 51 25" src="https://user-images.githubusercontent.com/639669/77969525-2cef1080-72ea-11ea-8aff-57db5300b401.png">

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.4.0/Message%20Queue.md

BUGS & IMPROVEMENTS:

- [X] (1) Target folder remove while uploading https://github.com/owncloud/ios-app/pull/662#issuecomment-652482467 [FIXED]
- [x] (2) Badge in app icon remains after reinstalling https://github.com/owncloud/ios-app/pull/662#issuecomment-652489908 [FIXED]
- [X] (3) Folder created in Files app is not removed even when was cancelled in the app https://github.com/owncloud/ios-app/pull/662#issuecomment-652534684 [MOVED TO https://github.com/owncloud/ios-app/issues/746]
- [x] (4) Notification points to other account https://github.com/owncloud/ios-app/pull/662#issuecomment-652976836
- [x] (5) https://github.com/owncloud/ios-app/issues/585 [FIXED]
- [X] (6) Keep both in quota oversize https://github.com/owncloud/ios-app/pull/662#issuecomment-654221845 [FIXED]
- [X] (7) Grouped errors https://github.com/owncloud/ios-app/pull/662#issuecomment-655989569 [DONE]

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=662) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=662) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: hosy<br/>:x: Michael Neuwert<br/>:x: felix-schwarz<br/><hr/>**Michael Neuwert** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=662) it.</sub>

hosy commented:
@felix-schwarz thank you for your good ideas! I think this will solve a lot of problems.

What do you think to make the Dismissable card in-app a little bit more inconspicuous? More like a notification, which will automatically dismiss, without an user interaction? Because the present implementation will block the UI flow too.

It is really cool that same issues will be grouped!

felix-schwarz commented:
@hosy Thanks for the feedback.

I think that typically, you'd get an error as a direct consequence of an action - and you'd then likely want to get a chance to act on it immediately. From that angle, the card approach has a number of benefits over notifications:
- it's positioned at the bottom of the screen, which users can reach easily
- it can be dismissed (to be dealt with later in the Status tab) by tapping in the area above it
- it can't be missed when the user isn't constantly looking at the screen (thinking of up-/downloads that fail at some point)
- no extra taps (taps on "Status", find message, etc…) needed to act on the message

However, we have full flexibility there - and I can see how - for cases with many errors in a short time frame - having card after card appear is not a good UX.

So, what we could do is to have the first message show up as card - and then, for as long as the card is visible, show all other messages only as normal iOS notifications. That'd ensure users can act on a single message immediately, but they don't get overwhelmed if a lot of issues occur in short order. What do you think?

hosy commented on ownCloud/Messages/AlertView.swift:
```diff
@@ -0,0 +1,217 @@
+			descriptionLabel.rightAnchor.constraint(equalTo: enclosure.rightAnchor, constant: -20),
+
+			optionStackView.leftAnchor.constraint(equalTo: enclosure.leftAnchor, constant: 20),
+			optionStackView.rightAnchor.constraint(equalTo: enclosure.rightAnchor, constant: -20)
```
Would be better to use constants for all these values and font sizes!?

felix-schwarz replied:
Good point! Changed that.

hosy commented on ownCloud/Messages/MessageGroupCell.swift:
```diff
@@ -0,0 +1,248 @@
+				if showAllLabel == nil {
+					showAllLabel = UIButton(type: .system)
+					showAllLabel?.translatesAutoresizingMaskIntoConstraints = false
+					showAllLabel?.setTitle("Show all", for: .normal)
```
`localize` is missing, be sure this string is available in localization file

felix-schwarz replied:
Good catch! Fixed.

hosy commented on ownCloud/Messages/MessageGroupCell.swift:
```diff
@@ -0,0 +1,248 @@
+					containerView.leftAnchor.constraint(equalTo: self.leftAnchor, constant: 20),
+					containerView.rightAnchor.constraint(equalTo: self.rightAnchor, constant: -20),
+					containerView.topAnchor.constraint(equalTo: self.topAnchor, constant: 20),
+					containerView.bottomAnchor.constraint(equalTo: self.bottomAnchor, constant: noBottomSpacing ? 0 : -20),
```
Please use constants instead of the values in this view

felix-schwarz replied:
Changed the code accordingly. Thanks!

hosy commented on ownCloud/Messages/MessageGroupCell.swift:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	@objc func applyAllSwitchChanged() {
+		badgeLabel?.labelText = "+ \((messageGroup?.messages.count ?? 0) - 1) more"
```
`localize` is missing (please check strings for localization)

felix-schwarz replied:
Thanks! Fixed.

hosy commented on ownCloud/Messages/MessageQueueExample.swift:
```diff
@@ -0,0 +1,87 @@
+import ownCloudSDK
+
+/*
+		// Example of using this example:
```
Maybe better: `// Example of using the MessageQueueExample `

felix-schwarz replied:
Good idea. Changed that.

hosy commented on ownCloud/Resources/en.lproj/Localizable.strings:
```diff
@@ -264,6 +264,8 @@
 "Forbidden Characters" = "Forbidden Characters";
 "File name cannot contain / or \\" = "File name cannot contain / or \\";
+"Item with same name already exists" = "Item with same name already exists";
+"An item with the same name already exists in this location." = "An item with the same name already exists in this location.";
```
This items are missing:
Apply choice to all similar issues
Show all

Please check if all strings was added

felix-schwarz replied:
Thanks. Checked and "Apply to all" was also missing. Fixed all of them.

hosy commented on ownCloud/Server List/ServerListTableViewController.swift:
```diff
@@ -102,6 +102,56 @@ class ServerListTableViewController: UITableViewController, Themeable {
+
+	var messageCountSelector : MessageSelector?
+
+	static let BookmarkMessageCountChanged = NSNotification.Name("boomark.message-count.changed")
```
Should you add a constant for this value: `"boomark.message-count.changed"`
Please check all other files, where this value was used.

felix-schwarz replied:
Moved this to its own `NSNotification.Name` extension now. Thanks.

felix-schwarz commented:
@hosy Thanks for the review. Addressed all findings as of https://github.com/owncloud/ios-app/pull/662/commits/1fae60df8eddb26bd92c22388a23e8fad03f359e.

hosy commented:
@felix-schwarz please can you check if all new buttons and actions are accessible with keyboard shortcuts? If commands are missing these should be added that everything could be reached with the keyboard. Thanks!

jesmrec commented:
### (1)

this is a corner case

1. Select some big files to upload (so, you'll have time to do the following steps)
2. Using another client, remove the target folder once the uploads start

Current: notifications enter in an endless loop. Sometimes, in full screen:

![Jul-01-2020 17-15-08](https://user-images.githubusercontent.com/14894746/86261429-d500c800-bbbe-11ea-9284-f4ddff302b2e.gif)

iPhoneXR, iOS13.4
commit `92e0103`

jesmrec commented:
### (2)

After executing (1), i had to remove the app from the device. After reinstalling, the badge in the app icon with the number of fails remain (status and list of bookmarks are clean). A fresh install should remove all the stuff

![ezgif com-optimize (7)](https://user-images.githubusercontent.com/14894746/86262883-bbf91680-bbc0-11ea-96d9-798ff2b2371e.gif)

If logs or whatever is needed here, let me know

iPhoneXR, iOS13.4
commit `92e0103`

jesmrec commented:
### (3)

(not totally sure this is related the current feature)

1. Open Files app
2. Create a folder with the same name as an existing file -> notification pops up
3. Click notification -> redirected to the app
4. Solve problem in the app (folder is cancelled)
5. Switch again to Files app

Current: Folder created en 2. and cancelled in 5. is not removed from Files App:

![ezgif com-optimize (8)](https://user-images.githubusercontent.com/14894746/86270846-2fecec00-bbcc-11ea-9470-62205d4c9c1b.gif)

Expected: Removed from Files App as well as in the app

iPhoneXR, iOS13.4
commit `92e0103`

felix-schwarz commented:
> @felix-schwarz please can you check if all new buttons and actions are accessible with keyboard shortcuts? If commands are missing these should be added that everything could be reached with the keyboard. Thanks!

For the "modal" alerts, I've added support for keyboard shortcuts with https://github.com/owncloud/ios-app/pull/662/commits/94233de1f4041acc086c4bd5b09551ab0890130e:

<img width="1320" alt="Bildschirmfoto 2020-07-02 um 12 04 55" src="https://user-images.githubusercontent.com/639669/86346060-aafde300-bc5c-11ea-82df-e325ce7a63b8.png">

jesmrec commented:
### (4) [IMPROVEMENT]


1. Attach two accounts to the app, and open one of them
2. Switch to Files App, and open the other account
3. Perform an action that causes an error (move a file to a folder with an existing name inside, create a folder with a name of a file...)
4. Notification pops up -> click on it

Current: Redirected to the app, but not to the account in which the error happened, but to the other one.
Expected: It'd be better to open the account in which the problem happened. 

iPhoneXR, iOS13.4
commit `92e0103`

jesmrec commented:
### (5)

about the issue attached in the first message: https://github.com/owncloud/ios-app/issues/585

Following the detailed steps: 

1. Login with OAuth2
2. Start to upload a bunch of files (f.ex, an album with 50 videos)
3. While it is being uploaded, remove token (webUI -> Settings -> Security - OAuth2 -> remove iOS app as authorized)
4. Get a new token by signing in

The error is still "ugly" for users, but presented in notification style:

![Screen Shot 2020-07-06 at 14 47 55](https://user-images.githubusercontent.com/14894746/86595012-5c23b680-bf98-11ea-89ae-bd785a2f72ca.png)

On the other side, the only way to get the "Sing in" dialog is reopening the app. 

iPhoneXR, iOS13.4
commit `92e0103`

jesmrec commented:
### (6)

1. Upload a file in order to overpass the account quota

Current: This is the notification:

![Uploading Screenshot 2020-07-06 at 15.01.34.png…]()

`Keep Both` option: not sure if this make sense...

Current: If you don't have enough space in your account, the only thing you can do is `Cancel`

iPhoneXR, iOS13.4
commit `92e0103`

felix-schwarz commented:
@jesmrec Going through the issues:

**(1)** I added an option to `Cancel` the upload with the latest commits.

**(2)** I think this is a bug in iOS. In the latest commit, I made a change so the badge count is set to 0 on start and then to the actual value - if that value is different. That should address the issue.

**(3)** This isn't related to the message queue, but an interesting find regardless.

**(4)** I had different thoughts when implementing this:
- if an error occurs in account A, but account B is the only one that's currently on screen (in split view for example), the tapped notification will bring up the message in account B, but with a title indicating the message belongs to account A.
- if an error occurs in account A, with account A and B open in the app, it'll try to show it in account A, but fall back to account B if it's the only one that's visible (again, with an additional title). Background here is that AFAIK iOS is allowed to close backgrounded "windows" at any time.
- leaving whatever other account a user is currently in and logging into the account the issue occurred is something I had implemented intermittently, but decided against, because it was a really confusing user experience and easy to loose a feeling of place for me as a user

**(5)** Thanks. I'll check this one tomorrow.

**(6)** `Keep both` here would add a number to the file to upload(`Photo.jpg`, `Photo (2).jpg`, …) so the existing and uploaded file can exist alongside. What should work different here?

felix-schwarz commented:
@jesmrec The latest commit addresses **(5)**. The error message now reads: `Couldn't upload … (error 3, oauth2: invalid_grant)`.

jesmrec commented:
(1) -> Makes much more sense to have the `Cancel` option. Then, other question comes to my mind: the number of problems to fix is limited to the number of maximum concurrent actions (uploads in that case). So, in case there are a large number of problems to solve (failed uploads), user will not be able to fix at the time. The option `Apply to all` means really, `Apply to all that are being processed/not queued`.  It might lead to confusion, and in case the number is really huge, it will really annoy. Is there something we can do to improve this? probably this is not in the scope of this PR, we can address in separate feature if there are other different stuff involved.

(2) -> fixed.

(3) -> i will open a separate issue

(4) ->

> if an error occurs in account A, but account B is the only one that's currently on screen (in split view for example), the tapped notification will bring up the message in account B, but with a title indicating the message belongs to account A.

i do not get this. Notification pops up in Files app, but the message does not indicate which account the problem happens:

![Screen Shot 2020-07-07 at 10 47 17](https://user-images.githubusercontent.com/14894746/86753870-b8daac00-c040-11ea-8b94-4ed0c1db53d7.png)

do i miss something?

Other two points are OK for me.

(6) -> 

> (6) Keep both here would add a number to the file to upload(Photo.jpg, Photo (2).jpg, …) so the existing and uploaded file can exist alongside. What should work different here?

There is no other file with the same name in the account that collides with the uploading file. That error is displayed when the account quota is exhausted. In such case, you are not allowed to upload and the only posible action is `Cancel`.

jesmrec commented:
Regadring this: https://github.com/owncloud/ios-app/pull/662#issuecomment-652916362

you mean, the notifications that pops from the bottom of the screen. For the actions in the `Status` view, there are no keyboard shortcuts available, aren't there?

felix-schwarz commented:
> (1) -> Makes much more sense to have the `Cancel` option. Then, other question comes to my mind: the number of problems to fix is limited to the number of maximum concurrent actions (uploads in that case). So, in case there are a large number of problems to solve (failed uploads), user will not be able to fix at the time. The option `Apply to all` means really, `Apply to all that are being processed/not queued`. It might lead to confusion, and in case the number is really huge, it will really annoy. Is there something we can do to improve this? probably this is not in the scope of this PR, we can address in separate feature if there are other different stuff involved.

That's a good point. The Sync Engine places actions into "budgets", to make sure it doesn't take on too much work at once (this is separate from the placement of actions into "lanes", which makes sure actions are performed in a non-conflicting order).

Until the latest commit, an action waiting for user interaction reduced the available budget, at one point lowering it to zero, so other actions had to wait. With the latest commit, actions waiting for user interaction no longer use any budget, so that f.ex. queued uploads will all be tried and - if they fail - their error messages can be dismissed together.

> (3) -> i will open a separate issue

Thanks. I've spent quite some time trying to find out what's happening here, but couldn't find anything we could or should do different.

FWIW the File Provider and app correctly notify the Files app of changes, the Files app requests a new file list - and the file list is returned without the item - but the Files app still displays it. At least for the moment, I therefore believe this to be a bug in the Files app. But it's still good to track this with an issue.

> (4) ->
> 
> > if an error occurs in account A, but account B is the only one that's currently on screen (in split view for example), the tapped notification will bring up the message in account B, but with a title indicating the message belongs to account A.
> 
> i do not get this. Notification pops up in Files app, but the message does not indicate which account the problem happens:

Thanks & good catch! I was thinking only about the message coming up in the app itself.

With the latest commit, notifications now include a reference to the account, if the user has more than one account

> (6) ->
> 
> > (6) Keep both here would add a number to the file to upload(Photo.jpg, Photo (2).jpg, …) so the existing and uploaded file can exist alongside. What should work different here?
> 
> There is no other file with the same name in the account that collides with the uploading file. That error is displayed when the account quota is exhausted. In such case, you are not allowed to upload and the only posible action is `Cancel`.
> 
> keep QAing this.

Thanks for pointing this out. Internally, a wrong error code had been returned, it got the same treatment as "item already exists". Fixed that and now you should get "Cancel" and "Retry".

> Regadring this: [#662 (comment)](https://github.com/owncloud/ios-app/pull/662#issuecomment-652916362)
> 
> you mean, the notifications that pops from the bottom of the screen. For the actions in the `Status` view, there are no keyboard shortcuts available, aren't there?

Correct. For these, no shortcuts are available because there can be any number of issues in the status table, so it'd not be clear which one a shortcut is targeting.

jesmrec commented:
(1) -> fixed.

(3) -> https://github.com/owncloud/ios-app/issues/746.

(4) -> fixed.

(6) -> fixed. I saw a small glitch when operation is retried. Warning and progress bar overlaps:

![Jul-08-2020 18-09-57](https://user-images.githubusercontent.com/14894746/86943190-6b486700-c146-11ea-9ef7-abd6b160c1c1.gif)

Starting the 2nd round.

jesmrec commented:
### (7)

The warning icon  ![Screenshot 2020-07-09 at 10 28 52](https://user-images.githubusercontent.com/14894746/87016487-0dac2d00-c1cf-11ea-9d29-d4fcdfe00ed9.png)  is not working on iPad.

in Phone: by clicking the warning icon, a card is displayed on the bottom to solve the problem individually
in iPad: nothing happens

iPhoneXR v13.4
iPadAir v13.4

felix-schwarz commented:
> (3) -> #746.

Thanks!

> (6) -> fixed. I saw a small glitch when operation is retried. Warning and progress bar overlap:
> 
> ![Jul-08-2020 18-09-57](https://user-images.githubusercontent.com/14894746/86943190-6b486700-c146-11ea-9ef7-abd6b160c1c1.gif)

I'm not sure what's happening here because the code that controls visibility of the progress ring and the alert button should mutually exclude that both are shown at once:

```swift
statusCircle.isHidden = hasMessage
messageButton.isHidden = !hasMessage
```

Is there any way I can reproduce this?

felix-schwarz commented:
Re **(7)**: it works almost like this:

iOS notifications don't support providing choices like `Retry` or `Cancel` directly with the notification. Instead, iOS wants apps to register "categories" and provide the choices with these. And then set the category's identifier for the iOS notification that should have these choices.

To support this, the SDK allows the registration of `OCMessageTemplate`s, from which this info can be derived, but which are also used to simplify the creation of error messages and avoid duplicate code.

Each of those `OCMessageTemplate`s has an identifier of its own, which is currently what is being used to group messages together:

<img width="879" alt="Bildschirmfoto 2020-07-10 um 22 41 44" src="https://user-images.githubusercontent.com/639669/87201287-8b228b00-c2fe-11ea-999f-90a6f038a9a0.png">

In the cases you described, the upload action uses the same template for uploads that failed due to a lack of space as well as those that failed because the upload targeted a non-existent directory … so these messages are grouped together.

If this should be a problem going forward, I can adapt the code to group each of these related but - at the root - different issues separately.

@jesmrec

jesmrec commented:
(6) is fixed
(7) clarified, and open to iterate. But the current solution is OK for me.

This is approved
